File: src/main/java/corp/soyummy/config/CustomErrorService.java
package corp.soyummy.config;

import org.springframework.stereotype.Service;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.security.web.access.AccessDeniedHandler;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.MediaType;

@Service
public class CustomErrorService {
    public AuthenticationEntryPoint customAuthenticationEntryPoint() {
        return (request, response, authException) -> {
            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
            response.setContentType(MediaType.APPLICATION_JSON_VALUE);
            response.getWriter().write(
                    "{ \"status\": 401, \"error\": \"Unauthorized\", \"message\": \"Invalid name or password\" }"
            );
        };
    }

    public AccessDeniedHandler customAccessDeniedHandler() {
        return (request, response, accessDeniedException) -> {
            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
            response.setContentType(MediaType.APPLICATION_JSON_VALUE);
            response.getWriter().write(
                    "{ \"status\": 403, \"error\": \"Access Denied\", \"message\": \"" + accessDeniedException.getMessage() + "\" }"
            );
        };
    }
}

File: src/main/kotlin/com/soyummy/config/CustomErrorService.kt
package com.soyummy.config

import org.springframework.stereotype.Service
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.security.web.access.AccessDeniedHandler
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.MediaType

@Service
class CustomErrorService {
    fun unauthorizedHandler(): AuthenticationEntryPoint {
        return AuthenticationEntryPoint { request, response, authException ->
            response.status = HttpServletResponse.SC_UNAUTHORIZED
            response.contentType = MediaType.APPLICATION_JSON_VALUE
            response.writer.write(
                """
                {
                    "status": 401,
                    "error": "Unauthorized",
                    "message": "Invalid name or password"
                }
                """.trimIndent()
            )
        }
    }

    fun accessDeniedHandler(): AccessDeniedHandler {
        return AccessDeniedHandler { request, response, accessDeniedException ->
            response.status = HttpServletResponse.SC_FORBIDDEN
            response.contentType = MediaType.APPLICATION_JSON_VALUE
            response.writer.write(
                """
                {
                    "status": 403,
                    "error": "Access Denied",
                    "message": "${accessDeniedException.message}"
                }
                """.trimIndent()
            )
        }
    }
}